Improve copy and attempt to clarify mechanism (#2)
ko1/nakayoshi_fork: README.md

## Usage

You only need to `require "nakayoshi_fork"` to enable the gem.

If you want to disable `nakayoshi_fork`, use `fork(nakayoshi: false)` or `fork(cow_friendly: false)`.

## Mechanism

### Ruby 2.1 GC

MRI 2.1 uses a generational GC algorithm that separate *young* objects and *old* objects:

1. *Young* objects are promoted to *old* objects when they survive 1 GC cycle.
2. After surviving 1 GC cycle, the `old` bit is set for each *young* object so we can tell that the promoted objects are now *old* objects.

MRI 2.1 also uses a bitmap to store the old object bit which means it's CoW friendly.

### Ruby 2.2 GC

MRI Ruby 2.2 uses an algorithm to promote *young* objects after 3 GC cycles instead of 1.

> To count GC survivors, all objects have an *age* field (2 bits field to count fron 0 to 3).

Reviewer: ```suggestion
To count GC survivors, all objects have an *age* field (2 bits field to count from 0 to 3).
```

Newly created *young* objects are age 0. Promoted *old* objects are age 3.

Unfortunately, the `age` fields are embedded into object headers. This means that when there 
are many young objects premature promotion is a problem because object headers are written 

> and marked as a dirty page.

Reviewer: "Premature" might not be the right word.  The issue is that if there are many young objects, and those objects get older, as they get older the GC will write to their age bits.  This is not great because if those young objects were created in the parent process, the writes done in the parent or child process will dirty the page, causing the page to be copied.

This is why MRI 2.2 has CoW friendly problem on fork.

> The `nakayoshi_fork` gem promotes most of young objects before fork by invoking GC some times.

Reviewer: ```suggestion
The `nakayoshi_fork` gem promotes most young objects before fork by invoking the GC a few times.
```

> The `nakayoshi_fork` gem promotes most of young objects before fork by invoking GC some times.

Reviewer: ```suggestion
The `nakayoshi_fork` gem promotes most young objects before fork by invoking the GC a few times.
```
